Introspected interface referencing Lombok builder class fails with java.lang.IllegalAccessError

### Expected Behavior

The interface should be successfully introspected at compile time, and an application context using the interface should start successfully.

### Actual Behaviour

With Micronaut 4.6.3, everything works fine. With Micronaut 4.7.6 it seems to fail either at compile time or at context initialization time, with an error.  An example stack trace:

```
Failed to load a service: java.lang.IllegalAccessError: class example.$Bar$Introspection tried to access protected method 'void java.lang.Record.<init>()' (example.$Bar$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
io.micronaut.core.io.service.SoftServiceLoader$ServiceLoadingException: Failed to load a service: java.lang.IllegalAccessError: class example.$Bar$Introspection tried to access protected method 'void java.lang.Record.<init>()' (example.$Bar$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
	at io.micronaut.core.io.service.ServiceScanner$ServiceInstanceLoader.collect(ServiceScanner.java:244)
	at io.micronaut.core.io.service.ServiceScanner$DefaultServiceCollector.collect(ServiceScanner.java:157)
	at io.micronaut.core.io.service.SoftServiceLoader.collectDynamicServices(SoftServiceLoader.java:199)
	at io.micronaut.core.io.service.SoftServiceLoader.collectAll(SoftServiceLoader.java:174)
	at io.micronaut.core.beans.DefaultBeanIntrospector.getIntrospections(DefaultBeanIntrospector.java:117)
	at io.micronaut.core.beans.DefaultBeanIntrospector.findIntrospection(DefaultBeanIntrospector.java:88)
	at io.micronaut.core.reflect.InstantiationUtils.tryInstantiate(InstantiationUtils.java:165)
	at io.micronaut.context.conditions.MatchesCustomCondition.matches(MatchesCustomCondition.java:58)
	at io.micronaut.context.AbstractInitializableBeanDefinitionAndReference.matches(AbstractInitializableBeanDefinitionAndReference.java:111)
	at io.micronaut.context.AbstractInitializableBeanDefinitionAndReference.isEnabled(AbstractInitializableBeanDefinitionAndReference.java:94)
	at io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.isDefinitionEnabled(DefaultBeanContext.java:4294)
	at io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.isDefinitionEnabled(DefaultBeanContext.java:4274)
	at io.micronaut.context.DefaultBeanContext.collectBeanCandidates(DefaultBeanContext.java:2167)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:2135)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidatesInternal(DefaultBeanContext.java:3429)
	at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3485)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1446)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:883)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:873)
	at io.micronaut.context.event.ApplicationEventPublisherFactory$2.lambda$$1(ApplicationEventPublisherFactory.java:215)
	at io.micronaut.core.util.SupplierUtil$1.get(SupplierUtil.java:47)
	at io.micronaut.context.event.ApplicationEventPublisherFactory$2.publishEvent(ApplicationEventPublisherFactory.java:226)
	at io.micronaut.context.DefaultBeanContext.publishEvent(DefaultBeanContext.java:1831)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:360)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:216)
	at io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:507)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:346)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeAll(MicronautJunit5Extension.java:84)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$13(ClassBasedTestDescriptor.java:396)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:396)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:212)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:85)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:153)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:146)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:144)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:100)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:160)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:146)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:144)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:100)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:124)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:94)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:92)
	at jdk.proxy1/jdk.proxy1.$Proxy4.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:200)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:132)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:103)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:63)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:121)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: io.micronaut.core.io.service.SoftServiceLoader$ServiceLoadingException: java.lang.IllegalAccessError: class example.$Bar$Introspection tried to access protected method 'void java.lang.Record.<init>()' (example.$Bar$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
	at io.micronaut.core.io.service.SoftServiceLoader.lambda$collectDynamicServices$2(SoftServiceLoader.java:195)
	at io.micronaut.core.io.service.ServiceScanner$ServiceInstanceLoader.compute(ServiceScanner.java:235)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.IllegalAccessError: class example.$Bar$Introspection tried to access protected method 'void java.lang.Record.<init>()' (example.$Bar$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
	at example.$Bar$Introspection.<clinit>(Unknown Source)
	at java.base/jdk.internal.misc.Unsafe.allocateInstance(Native Method)
	at java.base/java.lang.invoke.DirectMethodHandle.allocateInstance(DirectMethodHandle.java:520)
	at io.micronaut.core.io.service.SoftServiceLoader.lambda$collectDynamicServices$2(SoftServiceLoader.java:187)
	... 7 more
```

### Steps To Reproduce

1. create a class that implements an interface. Use Lombok Builder for the class, and add an annotation on the interface  that references the builder class.
2. compile
3. start a context using those classes

See the sample project attached. Run `./gradlew test --stacktrace` to see the issue. Note that with `micronautVersion=4.6.3`, the test passes. With `micronautVersion=4.7.6`, the test fails.

[micronaut-test-case.zip](https://github.com/user-attachments/files/18854276/micronaut-test-case.zip)

### Environment Information

Operating System: Windows 11 (but would happen with any)
Java 17

### Example Application

_No response_

### Version

4.7.6

--- a/core/src/main/java/io/micronaut/core/annotation/AnnotationClassValue.java
+++ b/core/src/main/java/io/micronaut/core/annotation/AnnotationClassValue.java
@@ -95,7 +95,7 @@ public final class AnnotationClassValue<T> implements CharSequence, Named {
         this.theClass = (Class<T>) instance.getClass();
         this.name = theClass.getName();
         this.instance = instance;
-        this.instantiated = true;
+        this.instantiated = !(instance instanceof AnnotationClassValue.UnresolvedClass);
     }
 
     /**
@@ -120,8 +120,7 @@ public final class AnnotationClassValue<T> implements CharSequence, Named {
      * @since 1.1
      */
     public boolean isInstantiated() {
-        return instantiated ||
-            (getInstance().isPresent() && !(getInstance().get() instanceof AnnotationClassValue.UnresolvedClass));
+        return instantiated;
     }
 
     /**

--- a/test-suite/src/test/java/io/micronaut/test/lombok/Bar.java
+++ b/test-suite/src/test/java/io/micronaut/test/lombok/Bar.java
@@ -1,0 +1,15 @@
+package io.micronaut.test.lombok;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import io.micronaut.core.annotation.Introspected;
+
+@Introspected
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonDeserialize(builder = Foo.FooBuilder.class)
+public interface Bar {
+
+    String getName();
+
+}
+

--- a/test-suite/src/test/java/io/micronaut/test/lombok/Foo.java
+++ b/test-suite/src/test/java/io/micronaut/test/lombok/Foo.java
@@ -1,0 +1,26 @@
+package io.micronaut.test.lombok;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import io.micronaut.core.annotation.Introspected;
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+import lombok.experimental.FieldNameConstants;
+import lombok.extern.jackson.Jacksonized;
+
+@Getter
+@Builder(toBuilder = true)
+@ToString
+@Jacksonized
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+@EqualsAndHashCode
+@FieldNameConstants(innerTypeName = "UserField", asEnum = true)
+@Introspected
+public class Foo implements Bar {
+
+    private String name;
+
+}

--- a/test-suite/src/test/java/io/micronaut/test/lombok/LombokIntrospectedBuilderTest.java
+++ b/test-suite/src/test/java/io/micronaut/test/lombok/LombokIntrospectedBuilderTest.java
@@ -16,6 +16,15 @@ import java.util.UUID;
 public class LombokIntrospectedBuilderTest {
 
     @Test
+    void testLombokInheritance() {
+        Foo foo = Foo.builder().name("Name").build();
+        final BeanIntrospection<Foo> introspection = BeanIntrospection.getIntrospection(Foo.class);
+
+        Assertions.assertEquals("Name", foo.getName());
+        Assertions.assertNotNull(introspection);
+    }
+
+    @Test
     void testNoArgsAndAllArgsConstructor() {
         BeanIntrospection<BarCommand> introspection = BeanIntrospection.getIntrospection(BarCommand.class);
 

